test/common/upstream/hds_test.cc

```diff
@@ -95,16 +100,72 @@ class HdsTest : public testing::Test {
         envoy::type::v3::HTTP1);
     health_check->mutable_health_checks(0)->mutable_http_health_check()->set_path("/healthcheck");
 
-    auto* socket_address = health_check->add_locality_endpoints()
-                               ->add_endpoints()
-                               ->mutable_address()
-                               ->mutable_socket_address();
+    auto* locality_endpoints = health_check->add_locality_endpoints();
+    // add locality information to this endpoint set of one endpoint.
+    auto* locality = locality_endpoints->mutable_locality();
+    locality->set_region("middle_earth");
+    locality->set_zone("shire");
+    locality->set_sub_zone("hobbiton");
+
+    // add one endpoint to this locality grouping.
+    auto* socket_address =
+        locality_endpoints->add_endpoints()->mutable_address()->mutable_socket_address();
     socket_address->set_address("127.0.0.0");
     socket_address->set_port_value(1234);
 
     return msg;
   }
 
+  // Creates a HealthCheckSpecifier message that contains several clusters, endpoints, localities,
+  // with only one health check type.
+  envoy::service::health::v3::HealthCheckSpecifier*
+  createComplexSpecifier(int n_clusters, int n_localities, int n_endpoints) {
+    // Final specifier to return.
+    envoy::service::health::v3::HealthCheckSpecifier* msg =
```

Comment:
Rather than returning raw pointer, best to switch to smart pointer, i.e. `unique_ptr`.

Reply:
Good point. For my new function, I now work on and return a unique_ptr unstead of holding the raw pointer.